src/basler_synced_cameras_node.cpp

```diff
@@ -0,0 +1,255 @@
+#include <ros/console.h>
+#include <ros/ros.h>
+
+#include <pylon/PylonIncludes.h>
+#include <pylon/gige/BaslerGigEInstantCamera.h>
+
+#include <XmlRpcValue.h>
+
+#include "basler_parameters.h"
+#include "image_publisher.h"
+
+std::vector<Pylon::CBaslerGigEInstantCamera> loadCameras(const XmlRpc::XmlRpcValue& cameras_yaml)
+{
+    Pylon::CTlFactory& tl_factory = Pylon::CTlFactory::GetInstance();
+    Pylon::DeviceInfoList_t devices;
+    ros::Rate r(0.5);
+
+    std::vector<Pylon::CBaslerGigEInstantCamera> cameras(cameras_yaml.size());
+    for (size_t index = 0; index < cameras_yaml.size(); ++index)
+    {
+        ROS_ASSERT_MSG(cameras_yaml[index].getType() == XmlRpc::XmlRpcValue::TypeStruct, "Bad yaml for camera %zu", index);
+        ROS_ASSERT_MSG(cameras_yaml[index].hasMember("serial"), "Camera %zu yaml missing serial", index);
+        std::string serial = std::to_string(int(cameras_yaml[index]["serial"]));
+
+        bool camera_found = false;
+        while(ros::ok() && !camera_found)
+        {
+            tl_factory.EnumerateDevices(devices);
+            for (size_t i = 0; i < devices.size(); ++i)
+            {
+                if (devices[i].GetSerialNumber().c_str() == serial)
+                {
+                    ROS_INFO_STREAM("Found camera " << index << " with matching serial " << serial);
+                    cameras[index].Attach(tl_factory.CreateDevice(devices[i]));
+                    camera_found = true;
+                    break;
+                }
+            }
+            if (!camera_found)
+            {
+                ROS_INFO_STREAM("Failed to find camera " << index << " with serial " << serial);
+                ROS_INFO_STREAM( devices.size() << " other serials found:");
+                for (size_t i = 0; i < devices.size(); ++i)
+                {
+                    ROS_INFO_STREAM(devices[i].GetSerialNumber());
+                }
+                r.sleep();
+                ros::spinOnce();
+            }
+        }
+    }
+    return cameras;
+}
+
+// Wait for all cameras to be slaves and to have same master clock id
+void waitForPTPSlave(const std::vector<Pylon::CBaslerGigEInstantCamera>& cameras)
+{
+    ROS_INFO("Waiting for %zu cameras to be PTP slaves...", cameras.size());
+    ros::Rate r(0.5);
+    while (ros::ok())
```

Comment:
You should have a timeout for this, too.  The default timeout can be pretty long (I know you said it might take a minute or two to get set up) but at some point, it should fail.  Ideally the overall timeout would be user-configurable.